go/exorelay/index.go

```diff
@@ -0,0 +1,143 @@
+package exorelay
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/Originate/exocom/go/exosocket"
+	"github.com/Originate/exocom/go/structs"
+)
+
+// ExoRelay is the mid level API to Exocom
+type ExoRelay struct {
+	closed               bool
+	exoSocket            *exosocket.ExoSocket
+	messageHandlers      MessageHandlerMapping
+	messageHandlersMutex sync.RWMutex
+	replyChannels        ReplyChannelMapping
+	replyChannelsMutex   sync.RWMutex
+}
+
+// NewExoRelay creates a new ExoRelay instance with the given config
+func NewExoRelay(config exosocket.Config) *ExoRelay {
+	return &ExoRelay{
+		closed:          false,
+		exoSocket:       exosocket.NewExoSocket(config),
+		replyChannels:   ReplyChannelMapping{},
+		messageHandlers: MessageHandlerMapping{},
+	}
+}
+
+// Connect connects to ExoCom and starts listening for messages
+func (e *ExoRelay) Connect() error {
+	err := e.exoSocket.Connect()
+	if err != nil {
+		return err
+	}
+	go e.listenForMessages()
+	return nil
+}
+
+// Close disconnects from ExoCom
+func (e *ExoRelay) Close() error {
+	if e.closed {
+		return nil
+	}
+	err := e.exoSocket.Close()
+	e.closed = true
+	return err
+}
+
+// RegisterHandler adds a handler for the given message name
+func (e *ExoRelay) RegisterHandler(messageName string, handler MessageHandler) {
+	e.messageHandlersMutex.Lock()
+	e.messageHandlers[messageName] = handler
+	e.messageHandlersMutex.Unlock()
+}
+
+// Send sends a message
+func (e *ExoRelay) Send(messageOptions exosocket.MessageOptions) (*structs.Message, error) {
+	return e.exoSocket.Send(messageOptions)
+}
+
+// WaitForActivity returns the first message received with the given activityID
+func (e *ExoRelay) WaitForActivity(activityID string, timeout time.Duration) (*structs.Message, error) {
+	messageChannel := make(chan *structs.Message)
+	e.addReplyChannel(activityID, messageChannel)
+	defer e.removeReplyChannel(activityID)
+	select {
+	case message := <-messageChannel:
+		return message, nil
+	case <-time.After(timeout):
+		return nil, fmt.Errorf("Did not receive a reply for '%s' after %s", activityID, timeout.String())
+	}
+}
+
+// Helpers
+
+func (e *ExoRelay) addReplyChannel(activityID string, messageChannel chan *structs.Message) {
+	e.replyChannelsMutex.Lock()
+	e.replyChannels[activityID] = messageChannel
```

Comment:
Can this accidentally override an already registered reply channel?

Reply:
Yes. I'll log that as a separate issue

Reply:
https://github.com/Originate/exocom/issues/275